--- a/configure.in
+++ b/configure.in
@@ -66,13 +66,23 @@
   *)
      AC_MSG_RESULT($withval)
-     if test -f $withval/pcap.h -a -f $withval/libpcap.a; then
+     if test -f $withval/pcap.h -o -f $withval/include/pcap.h -o -f $withval/include/pcap/pcap.h; then
         owd=`pwd`
         if cd $withval; then withval=`pwd`; cd $owd; fi
-	PCAP_CFLAGS="-I$withval -I$withval/bpf"
-	PCAPLIB="-L$withval -lpcap"
+	if test -f $withval/libpcap.a -o -f $withval/libpcap.so -o -f $withval/libpcap.so.*; then
+		PCAP_CFLAGS="-I$withval -I$withval/bpf"
+		PCAPLIB="-L$withval -lpcap"
+	elif test -f $withval/lib/libpcap.a -o -f $withval/lib/libpcap.so -o -f $withval/lib/libpcap.so.*; then
+		PCAP_CFLAGS="-I$withval/include -I$withval/include/pcap"
+		PCAPLIB="-L$withval/lib -lpcap"
+	else
+		AC_ERROR(libpcap not found in $withval)
+	fi
      else
-        AC_ERROR(pcap.h or libpcap.a not found in $withval)
+        AC_ERROR(pcap.h not found in $withval)
      fi
      ;;
   esac ],
